File: src/lib/api/services/addresses.ts
/**
 * Addresses API Service
 * 
 * User address management operations.
 */

import { apiClient } from "../client"

export type AddressType = "SHIPPING" | "BILLING" | "BOTH"

export interface UserAddress {
  id: string
  userId: string
  type: AddressType
  isDefault: boolean
  phone: string | null
  shippingAddress: string
  shippingCity: string
  shippingState: string | null
  shippingZip: string | null
  shippingCountry: string
  billingAddress: string | null
  billingCity: string | null
  billingState: string | null
  billingZip: string | null
  billingCountry: string | null
  label: string | null
  createdAt: string
  updatedAt: string
}

export interface UserAddressCreate {
  type?: AddressType
  is_default?: boolean
  phone?: string | null
  shipping_address: string
  shipping_city: string
  shipping_state?: string | null
  shipping_zip?: string | null
  shipping_country?: string
  billing_address?: string | null
  billing_city?: string | null
  billing_state?: string | null
  billing_zip?: string | null
  billing_country?: string | null
  label?: string | null
}

export interface UserAddressUpdate {
  type?: AddressType
  is_default?: boolean
  phone?: string | null
  shipping_address?: string
  shipping_city?: string
  shipping_state?: string | null
  shipping_zip?: string | null
  shipping_country?: string
  billing_address?: string | null
  billing_city?: string | null
  billing_state?: string | null
  billing_zip?: string | null
  billing_country?: string | null
  label?: string | null
}

export const addressesApi = {
  /**
   * Get all addresses for the current user
   */
  list: (): Promise<UserAddress[]> => {
    return apiClient.get<UserAddress[]>("/addresses")
  },

  /**
   * Get a specific address by ID
   */
  get: (id: string): Promise<UserAddress> => {
    return apiClient.get<UserAddress>(`/addresses/${id}`)
  },

  /**
   * Create a new address
   */
  create: (data: UserAddressCreate): Promise<UserAddress> => {
    return apiClient.post<UserAddress>("/addresses", data)
  },

  /**
   * Update an existing address
   */
  update: (id: string, data: UserAddressUpdate): Promise<UserAddress> => {
    return apiClient.patch<UserAddress>(`/addresses/${id}`, data)
  },

  /**
   * Delete an address
   */
  delete: (id: string): Promise<void> => {
    return apiClient.delete<void>(`/addresses/${id}`)
  },
}

File: backend/routers/addresses.py
"""
User Address Management Router
"""
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from prisma import Prisma
from prisma_client import get_prisma_client
from models.address_models import (
    UserAddressCreate,
    UserAddressUpdate,
    UserAddressResponse,
)
from routers.auth import get_current_user

router = APIRouter(prefix="/addresses", tags=["addresses"])


@router.get("", response_model=List[UserAddressResponse])
async def get_user_addresses(
    current_user=Depends(get_current_user),
    prisma: Prisma = Depends(get_prisma_client),
):
    """Get all addresses for the current user"""
    addresses = await prisma.useraddress.find_many(
        where={"userId": current_user.id},
        order={"createdAt": "desc"},
    )
    return addresses


@router.get("/{address_id}", response_model=UserAddressResponse)
async def get_address(
    address_id: str,
    current_user=Depends(get_current_user),
    prisma: Prisma = Depends(get_prisma_client),
):
    """Get a specific address by ID"""
    address = await prisma.useraddress.find_first(
        where={"id": address_id, "userId": current_user.id}
    )
    if not address:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Address not found",
        )
    return address


@router.post("", response_model=UserAddressResponse, status_code=status.HTTP_201_CREATED)
async def create_address(
    data: UserAddressCreate,
    current_user=Depends(get_current_user),
    prisma: Prisma = Depends(get_prisma_client),
):
    """Create a new address for the current user"""
    # If this is set as default, unset other defaults
    if data.is_default:
        await prisma.useraddress.update_many(
            where={"userId": current_user.id, "isDefault": True},
            data={"isDefault": False},
        )

    address = await prisma.useraddress.create(
        data={
            "userId": current_user.id,
            "type": data.type.value,
            "isDefault": data.is_default,
            "phone": data.phone,
            "shippingAddress": data.shipping_address,
            "shippingCity": data.shipping_city,
            "shippingState": data.shipping_state,
            "shippingZip": data.shipping_zip,
            "shippingCountry": data.shipping_country,
            "billingAddress": data.billing_address,
            "billingCity": data.billing_city,
            "billingState": data.billing_state,
            "billingZip": data.billing_zip,
            "billingCountry": data.billing_country,
            "label": data.label,
        }
    )
    return address


@router.patch("/{address_id}", response_model=UserAddressResponse)
async def update_address(
    address_id: str,
    data: UserAddressUpdate,
    current_user=Depends(get_current_user),
    prisma: Prisma = Depends(get_prisma_client),
):
    """Update an existing address"""
    # Verify ownership
    existing = await prisma.useraddress.find_first(
        where={"id": address_id, "userId": current_user.id}
    )
    if not existing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Address not found",
        )

    # If setting as default, unset other defaults
    update_data = {}
    if data.is_default is True:
        await prisma.useraddress.update_many(
            where={"userId": current_user.id, "isDefault": True},
            data={"isDefault": False},
        )
        update_data["isDefault"] = True
    elif data.is_default is False:
        update_data["isDefault"] = False

    # Build update data
    if data.type is not None:
        update_data["type"] = data.type.value
    if data.phone is not None:
        update_data["phone"] = data.phone
    if data.shipping_address is not None:
        update_data["shippingAddress"] = data.shipping_address
    if data.shipping_city is not None:
        update_data["shippingCity"] = data.shipping_city
    if data.shipping_state is not None:
        update_data["shippingState"] = data.shipping_state
    if data.shipping_zip is not None:
        update_data["shippingZip"] = data.shipping_zip
    if data.shipping_country is not None:
        update_data["shippingCountry"] = data.shipping_country
    if data.billing_address is not None:
        update_data["billingAddress"] = data.billing_address
    if data.billing_city is not None:
        update_data["billingCity"] = data.billing_city
    if data.billing_state is not None:
        update_data["billingState"] = data.billing_state
    if data.billing_zip is not None:
        update_data["billingZip"] = data.billing_zip
    if data.billing_country is not None:
        update_data["billingCountry"] = data.billing_country
    if data.label is not None:
        update_data["label"] = data.label

    address = await prisma.useraddress.update(
        where={"id": address_id},
        data=update_data,
    )
    return address


@router.delete("/{address_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_address(
    address_id: str,
    current_user=Depends(get_current_user),
    prisma: Prisma = Depends(get_prisma_client),
):
    """Delete an address"""
    # Verify ownership
    address = await prisma.useraddress.find_first(
        where={"id": address_id, "userId": current_user.id}
    )
    if not address:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Address not found",
        )

    await prisma.useraddress.delete(where={"id": address_id})
    return None
